### PR TITLE
Refactoring using XArray

### LINKED ISSUE #18: Error in the use of `marks_array2flags`
There is an error with respect to the dimensions

https://github.com/lina-usc/pylossless/blob/50f7e52f7176485ab2e028c9d900db961c226ac1/pylossless/pipeline.py#L542-L547

`flag_r_ch_inds` is actually an index along the epochs, whereas it should be along the channels. This create index-out-of-bound errors when an epochs with an index greater than the number of channels get flagged. I suspect that the same error is made across multiple use of `marks_array2flags`. I think that the use of numpy array is error prone. I suggest replacing them with XArrays which would allow to label axes and make operation less ambiguous. I also think we need to document, at a conceptual level, the different operations of the pipeline. I'll open a separate issue for this. 

Unfortunately, the file that triggered that error cannot be shared openly.

### PR DESCRIPTION
**Pull request by christian-oreilly**
Refactoring of the code to use more comprehensively XArray. This avoids making a series of operations across axes specified with numerical indices, which is opaque and confusing. Now using name dimensions and coordinates.

Fixes #18 

This PR also takes a big steps toward addressing #19 although some additional works need to be done to close that issue completely.

### CODE REVIEW

File: pylossless/pipeline.py [hunk 1]
```diff
@@ -111,7 +127,7 @@ def marks_array2flags(inarray, flag_dim='epoch', init_method='q', init_vals=(),
                       init_dir='both', init_crit=(), flag_method='z_score',
                       flag_vals=(), flag_crit=(), trim=0):
 
-    ''' This function takes an array typically created by chan_variance or
+    ''' This function takes an array with typically created by chan_variance or
```
Comment (scott-huberty) [inline pylossless/pipeline.py:130]: I think the new "with" here does not belong?

File: pylossless/pipeline.py [hunk 2]
```diff
@@ -550,7 +561,15 @@ def flag_ch_low_r(self, raw):
...
         # Edit the channel flag info structure
-        bad_ch_names = data_r_ch.ref_chan[flag_r_ch_inds].values.tolist()
+        try:
+            bad_ch_names = data_r_ch.ref_chan[flag_r_ch_inds].values.tolist()
+        except:
+            print(self.config['ch_low_r'])
+            print(data_r_ch.values.T)
+            print(data_r_ch.values.T.shape)
+            print(flag_r_ch_inds)
+            print(data_r_ch.ref_chan)
```
Comment (scott-huberty) [inline pylossless/pipeline.py:571]: Can we use logger here? I can loop back to this on another PR after merge

**Review by scott-huberty — APPROVED** [review]
Let some comments, one on the doc string and one on a print statement.

But the doc string needs to be refactored to follow numpy standards anyway, so I'll look back to it on another PR.

Ditto regarding the print, I'll integrate logger on a new PR.

File: pylossless/pipeline.py (hunk 1 above)
Comment (christian-oreilly) [reply pylossless/pipeline.py:130]: Yes, that may have been just due to the mess of rebasing and all. I did not intend to add the with. I probably removed it, and a commit due to rebasing put it back. It was a bit of a mess. Anyway, I'll squash all this commit when I merge.

File: pylossless/pipeline.py (hunk 2 above)
Comment (christian-oreilly) [reply pylossless/pipeline.py:571]: Sorry, I did not realize it had put you as a reviewer! This PR was actually not be meant to be revised at this time. I made it as a PR to more easily track its progress, but I still have to add many commits to this PR before it is ready for merging!

The print statements there, for example, are not meant to remain.

**Comment by christian-oreilly** [comment]
I am sure that there might be additional refactoring that will be done down the road, but the code of this branch is now working and fix #18. In the interest of avoiding divergence (and since their is no expectation of code stability yet), I recommend merging now. @scott-huberty are you OK with this?

**Comment by scott-huberty** [comment]
Yes. Merging now. Thanks @christian-oreilly !

**Comment by christian-oreilly** [comment]
Hey @scott-huberty Not a big problem, but just for the future: it is better for you to let me merge the PR I am leading. I had a few things I wanted to do before (e.g., last-minute small changes) and with the merging (squashing commits). I just needed you OK before moving along. I reverted the merge and I'll re-merge with these small changes.

**Comment by scott-huberty** [comment]
Oh sorry, I wasn't sure if I was the one that needed to merge!

**Comment by christian-oreilly** [comment]
Sure! No problem; just mentioning it for the future. I am currently running the version I have through a bunch of files... each one takes something like 1h to run (they are full recording; mostly long because of the ICA) so I am catching some bugs along the line and I want to include these within the PR before the merge. Should be done by the end of the weekend.

**Pull request merged**